File: app/src/main/java/com/matchthree/ui/game/StepPlayer.kt
package com.matchthree.ui.game

import androidx.compose.animation.core.AnimationSpec
import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.tween
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.snapshots.SnapshotStateMap
import androidx.compose.ui.geometry.Offset
import com.matchthree.game.engine.Step
import com.matchthree.game.model.Board
import com.matchthree.game.model.BoardConfig
import com.matchthree.game.model.Position
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlin.math.abs

/**
 * Plays back engine [Step] events as animations over the shared actor pool.
 *
 * The player owns the only mutable render state: a gem-id -> [GemActor] map plus
 * a logical id grid (which gem sits in which cell). The Canvas composable just
 * reads the actors; the ViewModel just resolves steps. This is the M2
 * incarnation of "UI plays back Steps" (AGENTS.md).
 *
 * Timing constants (from MECHANICS.md where they exist):
 *  - swap: ~150ms
 *  - invalid swap there-and-back: ~150ms
 *  - destroy/shrink: 200ms
 *  - falls/spawns: ~70ms per row, min 90ms (constant speed, simultaneous landing)
 */
class StepPlayer(
    private val config: BoardConfig = BoardConfig(),
    private val onSettled: (Board) -> Unit,
    private val onScore: (Int) -> Unit = {},
) {
    private val _actors = mutableStateMapOf<Int, GemActor>()

    /** Read-only view for the Canvas draw pass (redraws on change, no recomposition). */
    val actors: SnapshotStateMap<Int, GemActor> = _actors

    /** Logical gem-id grid in board-cell space; `null` = empty cell. */
    private var ids: Array<Array<Int?>> =
        Array(config.height) { arrayOfNulls(config.width) }

    fun hasGem(position: Position): Boolean =
        position.row in 0 until config.height &&
            position.col in 0 until config.width &&
            ids[position.row][position.col] != null

    /** Snaps the actor pool to a settled [board] (initial load and post-playback). */
    suspend fun applyBoard(board: Board) {
        val liveIds = board.positions().mapNotNull { board.gemAt(it)?.id }.toSet()
        _actors.keys.filterTo(mutableListOf()) { it !in liveIds }.forEach { id ->
            _actors.remove(id)
        }
        for (row in 0 until board.height) {
            for (col in 0 until board.width) {
                val gem = board.gemAt(row, col) ?: continue
                ids[row][col] = gem.id
                _actors[gem.id]?.snapTo(cellCenter(row, col))
                    ?: run { _actors[gem.id] = GemActor(gem.id, gem.type, cellCenter(row, col)) }
            }
        }
    }

    /** Plays a full engine resolution; invokes [onSettled] after the last step. */
    suspend fun play(steps: List<Step>) {
        steps.forEach { step -> playStep(step) }
        val settled = steps.filterIsInstance<Step.Settled>().lastOrNull()
        if (settled != null) onSettled(settled.board)
    }

    /** Animates an invalid swap: swap across and right back (no onSettled). */
    suspend fun playRejection(a: Position, b: Position) {
        swapActors(a, b)
        swapActors(a, b)
    }

    private suspend fun playStep(step: Step) {
        when (step) {
            is Step.Swap -> swapActors(step.a, step.b)
            is Step.Destroy -> destroyActors(step.positions)
            is Step.Score -> onScore(step.delta) // points accumulate live
            is Step.Fall -> fallActors(step.moves)
            is Step.Spawn -> spawnActors(step.gems)
            is Step.Settled -> Unit // handled in play()
        }
    }

    private suspend fun swapActors(a: Position, b: Position) {
        val idA = ids[a.row][a.col] ?: return
        val idB = ids[b.row][b.col] ?: return
        val actorA = _actors[idA] ?: return
        val actorB = _actors[idB] ?: return
        val targetA = cellCenter(b.row, b.col)
        val targetB = cellCenter(a.row, a.col)
        val spec = SWAP_SPEC
        coroutineScope {
            launch { actorA.moveTo(targetA, spec) }
            launch { actorB.moveTo(targetB, spec) }
        }
        ids[a.row][a.col] = idB
        ids[b.row][b.col] = idA
    }

    private suspend fun destroyActors(positions: Set<Position>) {
        val doomed = mutableListOf<GemActor>()
        for (p in positions) {
            val id = ids[p.row][p.col]
            if (id != null) {
                _actors[id]?.let { doomed += it }
                ids[p.row][p.col] = null
            }
        }
        if (doomed.isEmpty()) return
        coroutineScope {
            doomed.forEach { actor ->
                launch { actor.vanish(DESTROY_SPEC) }
            }
        }
        doomed.forEach { _actors.remove(it.id) }
    }

    private suspend fun fallActors(falls: List<Step.Fall.FallMove>) {
        if (falls.isEmpty()) return
        coroutineScope {
            falls.forEach { fall ->
                launch {
                    val actor = _actors[fall.gemId] ?: return@launch
                    val target = cellCenter(fall.to.row, fall.to.col)
                    val rows = abs(fall.to.row - fall.from.row)
                    actor.moveTo(target, fallSpec(rows))
                    ids[fall.to.row][fall.to.col] = fall.gemId
                    ids[fall.from.row][fall.from.col] = null
                }
            }
        }
    }

    private suspend fun spawnActors(spawned: List<Step.Spawn.Placement>) {
        if (spawned.isEmpty()) return
        // All gems start the same distance above their target so every column's
        // refill lands simultaneously at constant speed.
        val maxTargetRow = spawned.maxOf { it.position.row }
        val distance = maxTargetRow + 1
        coroutineScope {
            spawned.forEach { placement ->
                launch {
                    val gem = placement.gem
                    val target = cellCenter(placement.position.row, placement.position.col)
                    val start = Offset(target.x, target.y - distance)
                    val actor = GemActor(gem.id, gem.type, start)
                    _actors[gem.id] = actor
                    ids[placement.position.row][placement.position.col] = gem.id
                    actor.moveTo(target, fallSpec(distance))
                }
            }
        }
    }

    private fun cellCenter(row: Int, col: Int): Offset =
        Offset(col + 0.5f, row + 0.5f)

    private fun fallSpec(rows: Int): AnimationSpec<Float> =
        tween(durationMillis = FALL_BASE_MILLIS + rows * FALL_PER_ROW_MILLIS, easing = FastOutSlowInEasing)

    private companion object {
        const val SWAP_MILLIS = 150
        const val DESTROY_MILLIS = 200
        const val FALL_BASE_MILLIS = 90
        const val FALL_PER_ROW_MILLIS = 70

        val SWAP_SPEC = tween<Float>(SWAP_MILLIS, easing = FastOutSlowInEasing)
        val DESTROY_SPEC = tween<Float>(DESTROY_MILLIS, easing = FastOutSlowInEasing)
    }
}

File: View/StepPlayer.cs
using Godot;
using MatchThree.Engine.Model;
using MatchThree.Engine.Rules;

namespace MatchThree.View;

/// <summary>
/// Plays back engine <see cref="Step"/> events as animations over a pool of
/// instanced <see cref="GemActor"/> nodes.
///
/// The player owns the only mutable render state: a gem-id -&gt; actor map plus a
/// logical id grid (which gem sits in which cell). BoardView renders the nodes;
/// Game.cs resolves steps and hands them to the UI to play back (AGENTS.md).
///
/// Timing constants (from MECHANICS.md where they exist):
///  - swap: ~150ms
///  - invalid swap there-and-back: ~150ms
///  - destroy/shrink: 200ms
///  - falls/spawns: ~70ms per row, min 90ms (constant speed, simultaneous landing)
/// </summary>
public sealed class StepPlayer
{
    /// <summary>Swap duration (ms).</summary>
    public const float SwapMillis = 150f;

    /// <summary>Destroy/shrink duration (ms).</summary>
    public const float DestroyMillis = 200f;

    /// <summary>Base fall duration (ms).</summary>
    public const float FallBaseMillis = 90f;

    /// <summary>Additional duration per row fallen (ms).</summary>
    public const float FallPerRowMillis = 70f;

    private readonly Node _parent;
    private readonly BoardConfig _config;
    private readonly PackedScene _actorScene;
    private readonly float _cellSizePx;
    private readonly Dictionary<int, GemActor> _actors = new();
    private readonly int?[,] _ids;
    private readonly Action? _onSwap;
    private readonly Action<int>? _onDestroy;
    private int _destroyCount;

    /// <param name="parent">Node the actor instances are added to (the BoardView).</param>
    /// <param name="config">Board geometry (9x9).</param>
    /// <param name="cellSizePx">Pixel size of one board cell.</param>
    /// <param name="onSwap">Invoked when a genuine engine <see cref="Step.Swap"/>
    /// animates (accepted swap only; rejection playback stays silent).</param>
    /// <param name="onDestroy">Invoked when a <see cref="Step.Destroy"/> animates,
    /// with the 1-based cascade generation (the engine emits one Destroy step per
    /// cascade round, so this doubles as cascade depth for pitch rising).</param>
    public StepPlayer(Node parent, BoardConfig config, float cellSizePx, Action? onSwap = null, Action<int>? onDestroy = null)
    {
        _parent = parent;
        _config = config;
        _cellSizePx = cellSizePx;
        _ids = new int?[config.Height, config.Width];
        _onSwap = onSwap;
        _onDestroy = onDestroy;
        _actorScene = GD.Load<PackedScene>("res://Scenes/GemActor.tscn");
    }

    /// <summary>True when the cell currently holds a gem.</summary>
    public bool HasGem(Position position) =>
        position.Row >= 0 && position.Row < _config.Height &&
        position.Col >= 0 && position.Col < _config.Width &&
        _ids[position.Row, position.Col] is not null;

    /// <summary>Snaps the actor pool to a settled <paramref name="board"/> (initial load and post-playback).</summary>
    public void ApplyBoard(Board board)
    {
        var liveIds = board.Positions()
            .Select(p => board.GemAt(p))
            .Where(g => g is not null)
            .Select(g => g!.Value.Id)
            .ToHashSet();

        foreach (var id in _actors.Keys.Where(id => !liveIds.Contains(id)).ToList())
        {
            _actors[id].QueueFree();
            _actors.Remove(id);
        }

        for (var row = 0; row < board.Height; row++)
        {
            for (var col = 0; col < board.Width; col++)
            {
                var gem = board.GemAt(row, col);
                if (gem is null) continue;

                _ids[row, col] = gem.Value.Id;
                if (_actors.TryGetValue(gem.Value.Id, out var existing))
                {
                    existing.Position = CellCenter(row, col);
                }
                else
                {
                    var actor = SpawnActor(gem.Value);
                    actor.Position = CellCenter(row, col);
                }
            }
        }
    }

    /// <summary>Plays a full engine resolution; invokes <paramref name="onSettled"/> after the last step.</summary>
    public async Task PlayAsync(List<Step> steps, Action<Board>? onSettled, Action<int>? onScore)
    {
        // One resolution = one swap's full cascade; the destroy counter (and
        // with it the pop pitch) resets per resolution.
        _destroyCount = 0;
        foreach (var step in steps)
        {
            await PlayStepAsync(step, onScore);
        }

        var settled = steps.OfType<Step.Settled>().LastOrDefault();
        if (settled is not null) onSettled?.Invoke(settled.Board);
    }

    /// <summary>Animates an invalid swap: swap across and right back (no onSettled).</summary>
    public async Task PlayRejectionAsync(Position a, Position b)
    {
        await SwapActorsAsync(a, b);
        await SwapActorsAsync(a, b);
    }

    private async Task PlayStepAsync(Step step, Action<int>? onScore)
    {
        switch (step)
        {
            case Step.Swap swap:
                _onSwap?.Invoke();
                await SwapActorsAsync(swap.A, swap.B);
                break;
            case Step.ComboActivate:
                break; // the following Destroy animates the clear
            case Step.SpecialBirth birth:
                MarkSpecial(birth.GemId, birth.Special);
                break;
            case Step.Destroy destroy:
                _destroyCount++;
                _onDestroy?.Invoke(_destroyCount);
                await DestroyActorsAsync(destroy.Positions);
                break;
            case Step.Score score:
                onScore?.Invoke(score.Delta); // points accumulate live
                break;
            case Step.Fall fall:
                await FallActorsAsync(fall.Moves);
                break;
            case Step.Spawn spawn:
                await SpawnActorsAsync(spawn.Gems);
                break;
            case Step.Settled:
                break; // handled in PlayAsync
        }
    }

    /// <summary>Marks an existing actor as a special after a birth step.</summary>
    private void MarkSpecial(int gemId, Special special)
    {
        if (_actors.TryGetValue(gemId, out var actor)) actor.SetSpecial(special);
    }

    private async Task SwapActorsAsync(Position a, Position b)
    {
        var idA = _ids[a.Row, a.Col];
        var idB = _ids[b.Row, b.Col];
        if (idA is null || idB is null) return;
        if (!_actors.TryGetValue(idA.Value, out var actorA) || !_actors.TryGetValue(idB.Value, out var actorB)) return;

        var targetA = CellCenter(b.Row, b.Col);
        var targetB = CellCenter(a.Row, a.Col);
        await Task.WhenAll(
            actorA.MoveToAsync(targetA, SwapMillis),
            actorB.MoveToAsync(targetB, SwapMillis));

        _ids[a.Row, a.Col] = idB;
        _ids[b.Row, b.Col] = idA;
    }

    private async Task DestroyActorsAsync(ISet<Position> positions)
    {
        var doomed = new List<GemActor>();
        foreach (var p in positions)
        {
            var id = _ids[p.Row, p.Col];
            if (id is not null)
            {
                if (_actors.TryGetValue(id.Value, out var actor)) doomed.Add(actor);
                _ids[p.Row, p.Col] = null;
            }
        }
        if (doomed.Count == 0) return;

        await Task.WhenAll(doomed.Select(a => a.VanishAsync(DestroyMillis)));

        foreach (var actor in doomed)
        {
            _actors.Remove(actor.GemId);
            actor.QueueFree();
        }
    }

    private async Task FallActorsAsync(List<Step.Fall.FallMove> moves)
    {
        if (moves.Count == 0) return;
        await Task.WhenAll(moves.Select(FallActorAsync));
    }

    private async Task FallActorAsync(Step.Fall.FallMove fall)
    {
        if (!_actors.TryGetValue(fall.GemId, out var actor)) return;

        var target = CellCenter(fall.To.Row, fall.To.Col);
        var rows = Math.Abs(fall.To.Row - fall.From.Row);
        await actor.MoveToAsync(target, FallSpec(rows));

        _ids[fall.To.Row, fall.To.Col] = fall.GemId;
        _ids[fall.From.Row, fall.From.Col] = null;
    }

    private async Task SpawnActorsAsync(List<Step.Spawn.Placement> spawned)
    {
        if (spawned.Count == 0) return;

        // All gems start the same distance above their target so every column's
        // refill lands simultaneously at constant speed.
        var maxTargetRow = spawned.Max(p => p.Position.Row);
        var distance = maxTargetRow + 1;

        await Task.WhenAll(spawned.Select(p => SpawnActorAsync(p.Gem, p.Position, distance)));
    }

    private async Task SpawnActorAsync(Gem gem, Position position, int distance)
    {
        var target = CellCenter(position.Row, position.Col);
        var start = target - new Vector2(0f, distance * _cellSizePx);

        var actor = SpawnActor(gem);
        actor.Position = start;
        _ids[position.Row, position.Col] = gem.Id;

        await actor.MoveToAsync(target, FallSpec(distance));
    }

    private GemActor SpawnActor(Gem gem)
    {
        var actor = _actorScene.Instantiate<GemActor>();
        _parent.AddChild(actor);
        actor.Configure(gem.Id, gem.Type, gem.Special, _cellSizePx);
        _actors[gem.Id] = actor;
        return actor;
    }

    private Vector2 CellCenter(int row, int col) =>
        new((col + 0.5f) * _cellSizePx, (row + 0.5f) * _cellSizePx);

    private float FallSpec(int rows) => FallBaseMillis + rows * FallPerRowMillis;
}